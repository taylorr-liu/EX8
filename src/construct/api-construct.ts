import { RestApi, LambdaIntegration, Resource, RestApiProps } from '@aws-cdk/aws-apigateway';
import { Function } from '@aws-cdk/aws-lambda';
import { Construct } from '@aws-cdk/core';

export interface ItemsAPIProps extends RestApiProps {
  /**
     * @CURDLambdaFunctions Functions for CURD unspecified items
     */
  generalLambdaFunctions?: CURDLambdaFunctions;
  /**
     * @CURDLambdaFunctions Functions for CURD specified items
     */
  specificLambdaFunctions?: CURDLambdaFunctions;
}

export interface CURDLambdaFunctions {
  /**
     * @Function Function for the GET method
     */
  GET?: Function;

  /**
     * @Function Function for the POST method
     */
  POST?: Function;

  /**
     * @Function Function for the PUT method
     */
  PUT?: Function;

  /**
     * @Function Function for the DELETE method
     */
  DELETE?: Function;
}

export class ItemsAPI extends RestApi {
  /**
     * @Resource The API Resource of general methods.
     */
  public readonly generalMethodEntry: Resource;

  /**
     * @Resource The API Resource of specific methods.
     */
  public readonly specificMethodEntry: Resource;

  /**
     * The API with two kinds of resource, general methods and specific methods.
     * @param scope
     * @param id
     * @param props
     */
  constructor(scope: Construct, id: string, props: ItemsAPIProps) {
    super(scope, id, props);
    this.generalMethodEntry = this.root.addResource('items');
    this.specificMethodEntry = this.generalMethodEntry.addResource('{id}');

    if (props.generalLambdaFunctions) {
      this.setGeneralLambdaFunctions(props.generalLambdaFunctions);
    }
    if (props.specificLambdaFunctions) {
      this.setSpecificLambdaFunctions(props.specificLambdaFunctions);
    }
  }

  /**
     * Set the methods for the API resource.
     * @param resource
     * @param lambdaFuncs
     */
  private setLambdaFunctions(resource: Resource, lambdaFuncs: CURDLambdaFunctions) {
    if (lambdaFuncs.GET) {
      resource.addMethod('GET', new LambdaIntegration(lambdaFuncs.GET));
    }
    if (lambdaFuncs.POST) {
      resource.addMethod('POST', new LambdaIntegration(lambdaFuncs.POST));
    }
    if (lambdaFuncs.PUT) {
      resource.addMethod('PUT', new LambdaIntegration(lambdaFuncs.PUT));
    }
    if (lambdaFuncs.DELETE) {
      resource.addMethod('DELETE', new LambdaIntegration(lambdaFuncs.DELETE));
    }
  }

  /**
     * Set the lambda functions for the general methods.
     * @param lambdaFuncs The lambda functions for the general methods.
     */
  public setGeneralLambdaFunctions(lambdaFuncs: CURDLambdaFunctions) {
    this.setLambdaFunctions(
      this.generalMethodEntry,
      lambdaFuncs,
    );
  }

  /**
     * Set the lambda functions for the specific methods.
     * @param lambdaFuncs The lambda functions for the general methods.
     */
  public setSpecificLambdaFunctions(lambdaFuncs: CURDLambdaFunctions) {
    this.setLambdaFunctions(
      this.specificMethodEntry,
      lambdaFuncs,
    );
  }
}