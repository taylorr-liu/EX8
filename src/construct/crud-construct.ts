import { join } from 'path';
import { Table } from '@aws-cdk/aws-dynamodb';
import { Role } from '@aws-cdk/aws-iam';
import { Function } from '@aws-cdk/aws-lambda';
import { NodejsFunction } from '@aws-cdk/aws-lambda-nodejs';
import { Construct } from '@aws-cdk/core';

export interface CURDConstructProps {
  /**
     * @Table table The DynamoDB table to CURD.
     */
  table: Table;
  /**
     * @Role tableRole The IAM role with the policy to access the previous DynamoBD
     */
  tableRole: Role;
}

/**
 * A construct having CURD lambda functions.
 */
export class CURDConstruct extends Construct {
  /**
     * @Function The lambda function to get a specific item.
     */
  public readonly getOne: Function;
  /**
     * @Function The lambda function to get all items.
     */
  public readonly getAll: Function;
  /**
     * @Function The lambda function to create one item.
     */
  public readonly createOne: Function;
  /**
     * @Function The lambda function to update a specific item.
     */
  public readonly updateOne: Function;
  /**
     * @Function The lambda function to delete a specific item.
     */
  public readonly deleteOne: Function;

  /**
     * A construct having CURD lambda functions.
     * @param scope
     * @param id
     * @param props The props containing the DynamoDB Table to access and the IAM Role with accessibility.
     */
  constructor(scope: Construct, id: string, props: CURDConstructProps) {
    super(scope, id);

    this.getOne = new NodejsFunction(this, 'getOneItemFunction', {
      entry: `${process.env.APPS_PATH}/data/getone/index.js`,
      handler: 'handler',
      depsLockFilePath: `${process.env.APPS_PATH}/../package-lock.json`,
      environment: {
        TABLE_NAME: props.table.tableName,
        PRIMARY_KEY: 'itemId',
      },
      role: props.tableRole,
    });

    this.getAll = new NodejsFunction(this, 'getAllItemsFunction', {
      entry: `${process.env.APPS_PATH}/data/getall/index.js`,
      handler: 'handler',
      depsLockFilePath: `${process.env.APPS_PATH}/../package-lock.json`,
      environment: {
        TABLE_NAME: props.table.tableName,
        PRIMARY_KEY: 'itemId',
      },
      role: props.tableRole,
    });

    this.createOne = new NodejsFunction(this, 'createItemFunction', {
      entry: `${process.env.APPS_PATH}/data/create/index.js`,
      handler: 'handler',
      depsLockFilePath: `${process.env.APPS_PATH}/../package-lock.json`,
      environment: {
        TABLE_NAME: props.table.tableName,
        PRIMARY_KEY: 'itemId',
      },
      role: props.tableRole,
    });

    this.updateOne = new NodejsFunction(this, 'updateItemFunction', {
      entry: `${process.env.APPS_PATH}/data/update/index.js`,
      handler: 'handler',
      depsLockFilePath: `${process.env.APPS_PATH}/../package-lock.json`,
      environment: {
        TABLE_NAME: props.table.tableName,
        PRIMARY_KEY: 'itemId',
      },
      role: props.tableRole,
    });

    this.deleteOne = new NodejsFunction(this, 'deleteItemFunction', {
      entry: `${process.env.APPS_PATH}/data/delete/index.js`,
      handler: 'handler',
      depsLockFilePath: `${process.env.APPS_PATH}/../package-lock.json`,
      environment: {
        TABLE_NAME: props.table.tableName,
        PRIMARY_KEY: 'itemId',
      },
      role: props.tableRole,
    });
  }
}