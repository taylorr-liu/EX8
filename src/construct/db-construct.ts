import { Table, AttributeType } from '@aws-cdk/aws-dynamodb';
import { PolicyStatement, Effect } from '@aws-cdk/aws-iam';
import { Construct, RemovalPolicy } from '@aws-cdk/core';

export interface ItemTableProps {
  tableName: string;
  primaryKeyName: string;
}

export class ItemTable extends Table {
  constructor(scope: Construct, id: string, props: ItemTableProps) {
    super(scope, id, {
      partitionKey: {
        name: props.primaryKeyName,
        type: AttributeType.STRING,
      },
      tableName: props.tableName,
      removalPolicy: RemovalPolicy.DESTROY,
    });
  }
  getAccessPolicy(): PolicyStatement {
    return new PolicyStatement({
      effect: Effect.ALLOW,
      actions: [
        'dynamodb:PutItem',
        'dynamodb:UpdateItem',
        'dynamodb:DeleteItem',
        'dynamodb:GetItem',
        'dynamodb:Query',
        'dynamodb:Scan',
      ],
      resources: [this.tableArn],
    });
  }
}