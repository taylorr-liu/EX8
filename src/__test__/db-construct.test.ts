import '@aws-cdk/assert/jest';
import { App, Stack } from '@aws-cdk/core';
import { ItemTable } from '../construct/db-construct';

test('db-construct', () => {
  const app = new App();
  const stack = new Stack(app, 'test');
  const table = new ItemTable(stack, 'test-ItemTable', {
    tableName: 'items',
    primaryKeyName: 'itemId',
  });
  const policy = table.getAccessPolicy();

  expect(stack).toHaveResource('AWS::DynamoDB::Table');
  expect(policy.hasResource).toBeTruthy();
  [
    'dynamodb:PutItem',
    'dynamodb:UpdateItem',
    'dynamodb:DeleteItem',
    'dynamodb:GetItem',
    'dynamodb:Query',
    'dynamodb:Scan',
  ].forEach((action) => {
    expect(policy.toJSON().Action).toContain(action);
  });
  expect(app.synth().getStackArtifact(stack.artifactId).template)
    .toMatchSnapshot();
});