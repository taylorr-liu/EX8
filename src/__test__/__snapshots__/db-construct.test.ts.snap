// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`db-construct 1`] = `
Object {
  "Resources": Object {
    "testItemTable0E29024D": Object {
      "DeletionPolicy": "Delete",
      "Properties": Object {
        "AttributeDefinitions": Array [
          Object {
            "AttributeName": "itemId",
            "AttributeType": "S",
          },
        ],
        "KeySchema": Array [
          Object {
            "AttributeName": "itemId",
            "KeyType": "HASH",
          },
        ],
        "ProvisionedThroughput": Object {
          "ReadCapacityUnits": 5,
          "WriteCapacityUnits": 5,
        },
        "TableName": "items",
      },
      "Type": "AWS::DynamoDB::Table",
      "UpdateReplacePolicy": "Delete",
    },
  },
}
`;
